src/Microsoft.ML.Data/Depricated/Vector/VBufferMathUtils.cs

```diff
@@ -81,12 +85,13 @@ public static Float Sum(in VBuffer<Float> a)
         /// <param name="c">Value to multiply vector with</param>
         public static void ScaleBy(ref VBuffer<Float> dst, Float c)
         {
-            if (c == 1 || dst.Count == 0)
+            if (c == 1 || dst.GetValues().Length == 0)
                 return;
```

Comment:
Note that checks like this were *intended* to be quick, that is, "in this trivial border case, do nothing." Now though, since this check actually entails a fair amount of effort (i.e., creating the span), it is no longer appropriate as this sort of "border case."

Reply:
That is, if you just remove this condition, I wonder what happens? I guess the critical issue is whether we guard against null pointers in this `CpuMathUtils.Scale` thing.

---
In reply to: [232434600](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232434600) [](ancestors = 232434600)

Reply:
We don't guard against null pointers in `CpuMathUtils` at all. They all have asserts that the Span is non-empty. [https://github.com/dotnet/machinelearning/blob/28fb3bbca4d5b8ad5b1f0bc28fef35db4d8232f3/src/Microsoft.ML.CpuMath/Sse.cs#L103](https://github.com/dotnet/machinelearning/blob/28fb3bbca4d5b8ad5b1f0bc28fef35db4d8232f3/src/Microsoft.ML.CpuMath/Sse.cs#L103)

---
In reply to: [232435262](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232435262) [](ancestors = 232435262,232434600)

Reply:
I wonder if we can get away with another property next to `Length`, something like `ExplicitValuesLength`? This seems to be quite a common pattern to start by computing the length, and then maybe doing nothing else at all

---
In reply to: [232823696](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232823696) [](ancestors = 232823696,232435262,232434600)

Reply:
I discussed this a bit in https://github.com/dotnet/machinelearning/issues/608#issuecomment-433185895.

I find it really confusing to have so many "length" properties/fields/etc.  We have `Length`, `GetValues().Length`, `GetIndices().Length` so far.  Adding another public "Length" doesn't make it better IMO.

It doesn't happen often where we just get the length of the values, but do nothing else with it.  We can change this code to always pass the empty span into the `CpuMathUtils` and have the CpuMathUtils handle an empty span, if you'd like.

Reply:
I did a quick search, and this was the only place that called into `CpuMathUtils` and needed to check the length of the values, but do nothing else with the `GetValues()` span.

Getting a span from an array is a fairly cheap operation (much cheaper than a bunch of multiplication that is about to happen). So I don't feel this one place justifies adding a new public property, or changing the invariant in `CpuMathUtils` that it needs to be called with a non-empty span.

---
In reply to: [233670997](https://github.com/dotnet/machinelearning/pull/1580#discussion_r233670997) [](ancestors = 233670997)